community/kernel/src/main/java/org/neo4j/kernel/impl/logging/AsyncLogProvider.java

```diff
@@ -0,0 +1,142 @@
+    {
+        this.scheduler = scheduler;
+        this.events = new AsyncEvents<>( new Consumer<AsyncLogEvent>()
+        {
+            private final CappedOperation<Throwable> errorNotification = new CappedOperation<Throwable>(
+                    differentItemClasses(), time( 1, SECONDS ) )
+            {
+                @Override
+                protected void triggered( Throwable error )
+                {
+                    errorHandler.receive( error );
+                }
+            };
+
+            @Override
+            public void accept( AsyncLogEvent logEvent )
+            {
+                try
+                {
+                    logEvent.process();
+                }
+                catch ( Throwable t )
+                {
+                    // We catch Throwable here otherwise there's a high chance an error, like OutOfMemory
+                    // or similar will halt the logging from that point and onwards, since all logging
+                    // is processed by a dedicated thread.
+                    //
+                    // What's most important is that logging gets going, but we also can be nice and...
+                    // well... not log, since apparently there was a problem doing that right now, but do something.
+                    // That something will be to print stack trace of the exception, capped by different
+                    // exception types and time interval for good measure.
+                    try
+                    {
+                        errorNotification.event( t );
+                    }
+                    catch ( Throwable e )
+                    {
+                        // Even that failed. There's no reasonable thing left to do here, is there?
+                    }
+                }
+            }
+        } );
+        this.actual = actual;
+    }
+
+    @Override
+    public Log getLog( @SuppressWarnings( "rawtypes" ) Class loggingClass )
+    {
+        return new AsyncLog( events, actual.getLog( loggingClass ) );
+    }
+
+    @Override
+    public Log getLog( String name )
+    {
+        return new AsyncLog( events, actual.getLog( name ) );
+    }
+
+    @Override
+    public void init() throws Throwable
+    {
+        scheduler.execute( events );
+    }
+
+    /**
+     * Closes the asynchronous aspect of this {@link LogProvider}, not the actual {@link LogProvider}
+     * underneath. Future logging calls made to this {@link LogProvider} will be processed directly
+     * in the thread making the call.
+     */
+    @Override
+    public void shutdown()
+    {
+        // Shut down events so that future events will not be queued, rather processed immediately.
+        events.shutdown();
+
+        // Flush the queue right after we've shut it down since we don't want to have pending logging
+        // events after leaving this method. Right after leaving this method the actual output stream
+        // will most likely be closed.
+        events.run();
```

Comment:
Imho this run should be part of `events` itself and not to be leaked to AsyncLogProvider.
Events should insure that all queued messages are flushed on shutdown, instead of AsyncLogProvider doing that.